### PR TITLE
Initial implementation of aws metadata filter

### PR DESCRIPTION
**Pull request by PettitWesley**
#1780 

### Phase 1: Basic EC2 Metadata

The initial implementation will simply add the EC2 instance ID and AZ to the logs. 

Fluent Bit Configuration:

```
[FILTER]
    Name aws
```

Log with metadata fields:

```
{
  "ec2_instance_id": "i-06bc83dbc2ac2fdf8",
  "az"            : "us-west-1b"
}
```

`"az"` is the field named used here: https://github.com/takus/fluent-plugin-ec2-metadata
`"ec2_instance_id"` is the field name used by FireLens.

### CODE REVIEW

**Comment by PettitWesley** [comment]
Currently when I attempt to use the plugin, Fluent Bit crashes. The segfault has something to do with pthreads that I am having trouble debugging. I'm guessing some how my new code is causing this problem... but its not clear since my code does not touch the variable that is unexpectedly null.

The problem is that the thread pointer returned here is NULL: https://github.com/fluent/fluent-bit/pull/1795/files#diff-470d2bb097805a2895a745728cc5acc0R133

This leads to a segfault when we reach this line: https://github.com/fluent/fluent-bit/pull/1795/files#diff-27823f83ec0786de22afc84f24716e01R145

Output:
```
[2019/12/08 05:58:31] [debug] [storage] [cio stream] new stream registered: dummy.0
[2019/12/08 05:58:31] [ info] [storage] initializing...
[2019/12/08 05:58:31] [ info] [storage] in-memory
[2019/12/08 05:58:31] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/12/08 05:58:31] [ info] [engine] started (pid=5922)
[2019/12/08 05:58:31] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2019/12/08 05:58:31] [debug] [debug_aws] in cb_aws_init
[2019/12/08 05:58:31] [debug] [router] match rule dummy.0:stdout.0
[2019/12/08 05:58:31] [ info] [sp] stream processor started
[2019/12/08 05:58:31] [debug] [debug_aws] in cb_aws_filter
[2019/12/08 05:58:31] [debug] [debug_aws] in get_ec2_metadata
[2019/12/08 05:58:31] [debug] 0x7f529b82b000
[2019/12/08 05:58:31] [debug] 0x7f529b845180
[2019/12/08 05:58:31] [debug] [debug_aws] problem is here:
[2019/12/08 05:58:31] [debug] [debug_aws] value of 'flb_thread_key': 2
[2019/12/08 05:58:31] [debug] [debug_aws] in create_conn
[2019/12/08 05:58:31] [debug] [debug_aws] value of pointer 'th': (nil)
[2019/12/08 05:58:31] [debug] [debug_aws] problem is here:
[engine] caught signal (SIGSEGV)
[2019/12/08 05:58:31] [debug] [debug_aws] value of pointer 'th' (nil)
#0  0x4d9f71            in  flb_thread_yield() at include/fluent-bit/flb_thread_libco.h:75
#1  0x4d9f71            in  flb_io_net_connect() at src/flb_io.c:145
#2  0x4c6899            in  create_conn() at src/flb_upstream.c:155
#3  0x4c6b16            in  flb_upstream_conn_get() at src/flb_upstream.c:234
#4  0x537f6f            in  get_ec2_metadata() at plugins/filter_aws/aws.c:205
#5  0x538161            in  cb_aws_filter() at plugins/filter_aws/aws.c:267
#6  0x4b9183            in  flb_filter_do() at src/flb_filter.c:117
#7  0x4d903e            in  flb_input_chunk_append_raw() at src/flb_input_chunk.c:461
#8  0x4f122f            in  in_dummy_collect() at plugins/in_dummy/in_dummy.c:66
#9  0x4b8c50            in  flb_input_collector_fd() at src/flb_input.c:862
#10 0x4c263c            in  flb_engine_handle_event() at src/flb_engine.c:262
#11 0x4c263c            in  flb_engine_start() at src/flb_engine.c:489
#12 0x43b632            in  main() at src/fluent-bit.c:856
#13 0x7f529e6b7029      in  ???() at ???:0
#14 0x439b79            in  ???() at ???:0
#15 0xffffffffffffffff  in  ???() at ???:0
Aborted
```

File: src/flb_io.c [hunk 1]
```diff
@@ -141,6 +140,8 @@ FLB_INLINE int flb_io_net_connect(struct flb_upstream_conn *u_conn,
          * Return the control to the parent caller, we need to wait for
          * the event loop to get back to us.
          */
+        flb_debug("[debug_aws] problem is here:");
+        flb_debug("[debug_aws] value of pointer 'th' %p", th);
         flb_thread_yield(th, FLB_FALSE);
```
Comment (PettitWesley) [inline src/flb_io.c]: Crash happens when we reach this line.

File: src/flb_upstream.c [hunk 2]
```diff
@@ -127,7 +127,13 @@ static struct flb_upstream_conn *create_conn(struct flb_upstream *u)
 {
     int ret;
     struct flb_upstream_conn *conn;
+    flb_debug("[debug_aws] problem is here:");
+    // Looks like pthread_key_t is an unsigned int
+    flb_debug("[debug_aws] value of 'flb_thread_key': %u", flb_thread_key);
     struct flb_thread *th = pthread_getspecific(flb_thread_key);
```
Comment (PettitWesley) [inline src/flb_upstream.c]: `th` is NULL here, which eventually leads to the crash.

**Comment by PettitWesley** [comment]
Fluent Bit Config file:

```
[SERVICE]
    Log_Level debug


[INPUT]
    Name dummy
    Tag dummy

[FILTER]
    Name aws
    Match *

[OUTPUT]
    Name stdout
    Match *
```

Output:

```
$ ./bin/fluent-bit -c ~/aws-metadata/fluent-bit.conf
Fluent Bit v1.4.0
Copyright (C) Treasure Data

[2019/12/10 07:46:57] [debug] [storage] [cio stream] new stream registered: dummy.0
[2019/12/10 07:46:57] [ info] [storage] initializing...
[2019/12/10 07:46:57] [ info] [storage] in-memory
[2019/12/10 07:46:57] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/12/10 07:46:57] [ info] [engine] started (pid=25542)
[2019/12/10 07:46:57] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2019/12/10 07:46:57] [debug] [router] match rule dummy.0:stdout.0
[2019/12/10 07:46:57] [ info] [sp] stream processor started
[2019/12/10 07:46:57] [debug] [filter_aws] IMDSv2 token request http_do=0, HTTP Status: 200
[2019/12/10 07:46:57] [debug] [filter_aws] IMDSv2 metadata request http_do=0, HTTP Status: 200
[2019/12/10 07:46:58] [debug] [filter_aws] IMDSv2 metadata request http_do=0, HTTP Status: 200
[0] dummy: [1575964018.000141277, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[1] dummy: [1575964019.000112593, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[2] dummy: [1575964020.000110409, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[3] dummy: [1575964021.000109840, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[2019/12/10 07:47:01] [debug] [task] created task=0x7fef20440140 id=0 OK
[2019/12/10 07:47:01] [debug] [task] destroy task=0x7fef20440140 (task_id=0)
```

File: plugins/filter_aws/aws.c [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+    if (ctx != NULL) {
+        flb_filter_aws_destroy(ctx);
+    }
+    return 0;
+}
+
+struct flb_filter_plugin filter_aws_plugin = {
+    .name         = "aws",
+    .description  = "inject aws metadata",
```
Comment (edsiper) [inline plugins/filter_aws/aws.c]: I suggest: Add AWS Metadata

**Comment by edsiper** [comment]
@PettitWesley 

hey!, thanks for the contribution, the overall code looks good to me, I added a couple of comments requiring some minor changes which are mostly about code styling.

File: plugins/filter_aws/aws.c [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        flb_free(ctx->imds_v2_token);
+    }
+
+    if (ctx->availability_zone) {
+        flb_free(ctx->availability_zone);
+    }
+
+    if (ctx->instance_id) {
+        flb_free(ctx->instance_id);
+    }
```
Comment (PettitWesley) [inline plugins/filter_aws/aws.c]: Ran valgrind... these are sds strings and need to use `flb_sds_destroy` instead

**Comment by PettitWesley** [comment]
Valgrind:

```
[2] dummy: [1576022041.000222475, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[3] dummy: [1576022042.000230903, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[4] dummy: [1576022043.000229302, {"message"=>"dummy", "az"=>"us-west-2b", "ec2_instance_id"=>"i-06bc83dbc2ac2fdf8"}]
[2019/12/10 23:54:04] [debug] [task] destroy task=0x9546d40 (task_id=0)
^C[engine] caught signal (SIGINT)
[2019/12/10 23:54:05] [ info] [input] pausing dummy.0
==5376==
==5376== HEAP SUMMARY:
==5376==     in use at exit: 0 bytes in 0 blocks
==5376==   total heap usage: 333 allocs, 333 frees, 1,734,139 bytes allocated
==5376==
==5376== All heap blocks were freed -- no leaks are possible
==5376==
==5376== For counts of detected and suppressed errors, rerun with: -v
==5376== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

File: plugins/filter_aws/aws.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+#define FLB_FILTER_AWS_INSTANCE_ID_KEY_LEN                15
+
+struct flb_filter_aws {
+    /* upstream connection to ec2 IMDS */
+    struct flb_upstream *ec2_upstream;
+
+    /* IMDSv2 requires a token which must be present in metadata requests
+    This plugin does not refresh the token */
+    flb_sds_t imds_v2_token;
+    size_t imds_v2_token_len;
```
Comment (PettitWesley) [inline plugins/filter_aws/aws.h]: Change to int

File: plugins/filter_aws/aws.c [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        if (client->resp.payload_size > 0) {
+            flb_debug("[filter_aws] IMDSv2 token response\n%s",
+                      client->resp.payload);
+        }
+        flb_http_client_destroy(client);
+        flb_upstream_conn_release(u_conn);
+        return -1;
+    }
+
+    ctx->imds_v2_token = flb_sds_create_len(client->resp.payload,
```
Comment (PettitWesley) [inline plugins/filter_aws/aws.c]: Check that memory was actually allocated

**Comment by vector-sec** [comment]
@PettitWesley I'm actively following this from #1780 as I'm super interested in this functionality.

Is there a reason why you're limiting the filter to only provide az and ec2_instance_id? Would it be possible to add some of the other properties that are available in https://github.com/takus/fluent-plugin-ec2-metadata such as account_id, ami_id, vpc_id, hostname, etc?

**Comment by PettitWesley** [comment]
@vector-sec I want to keep this initial PR simple.

I'm also concurrently working on Amazon ElasticSearch support (#1804). I personally think that is higher priority than this plugin- I've had a number of individuals tell me that the AWS integration with Fluent Bit is "useless" until we implement that feature. 

Once I've finished an initial implementation of that I can circle back to this feature and add more fields (and then circle back to Amazon ES because the initial implementation won't have everything).

I will add a comment in #1780 on my design for 'Phase 2'. Part of my plan is to set up the basic scaffolding for this plugin (and some others) so that it's then very easy for other engineers to join in and make small changes to them to add all the desired functionality.

**Comment by vector-sec** [comment]
@PettitWesley understood, thanks for the clarification and thank you for working on this feature to begin with!

File: plugins/filter_aws/aws.c [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+                                            client->resp.payload_size);
+    ctx->imds_v2_token_len = client->resp.payload_size;
+
+    flb_http_client_destroy(client);
+    flb_upstream_conn_release(u_conn);
+    return 0;
+}
+
+static int get_metadata(struct flb_filter_aws *ctx, char *metadata_path,
+                        flb_sds_t *metadata, unsigned int *metadata_len)
```
Comment (PettitWesley) [inline plugins/filter_aws/aws.c]: use `size_t` instead of `unsigned int` for response length

**Comment by PettitWesley** [comment]
Added an option to select which metadata server version is used:

```
[FILTER]
    Name aws
    Match *
    imds_version v1
```

```
[FILTER]
    Name aws
    Match *
    imds_version v2
```

Default is v2 since that's EC2's recommendation. Many users who run Fluent Bit in a container will need to set the option to use v1. I've tested out both; including running inside of Docker.

**Comment by PettitWesley** [comment]
@edsiper This is good to go for another round of review. If possible, I'd like to get this into v1.4

**Comment by PettitWesley** [comment]
Doc PR: https://github.com/fluent/fluent-bit-docs/pull/255

**Pull request merged**